### PR TITLE
DRILL-5940: Add support for Avro messages with schema registry for Kafka

### PR DESCRIPTION
**Pull request by vvysotskyi**
# [DRILL-5940](https://issues.apache.org/jira/browse/DRILL-5940): Add support for Avro messages with schema registry for Kafka

## Description
- Refactored `KafkaRecordReader` to use EVF.
- Updated `JsonMessageReader` to use V2 JSON reader
- Added `AvroMessageReader` for processing messages serialized as Avro records and added schema registry support.

To be able to deserialize Kafka messages of Avro type, schema registry URL should be specified in the Kafka storage plugin `kafkaConsumerProps` (and other schema registry options if required):
``` 
"schema.registry.url": "http://localhost:8081"
```
After that, `store.kafka.record.reader` session option should be set to `org.apache.drill.exec.store.kafka.decoders.AvroMessageReader`.

## Documentation
NA

## Testing
Added UT that reads Avro messages.

### CODE REVIEW

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java [hunk 1]
```diff
@@ -17,131 +17,105 @@
...
-        writer.setPosition(currentMessageCount);
-        boolean status = messageReader.readMessage(consumerRecord);
-        // increment record count only if message was read successfully
-        if (status) {
-          if (++currentMessageCount >= DEFAULT_MESSAGES_PER_BATCH) {
-            break;
-          }
-        }
+  public boolean next() {
+    RowSetLoader rowWriter = messageReader.getResultSetLoader().writer();
```
Comment (cgivre) [inline contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java:78]: @vvysotskyi 
Does this line belong in the `setup()` method?  In the Splunk plugin for example:

https://github.com/apache/drill/blob/bc53c8e6c4a24b7cbce2112690a7d3e77e55fe41/contrib/storage-splunk/src/main/java/org/apache/drill/exec/store/splunk/SplunkBatchReader.java#L122-L129

Then in each iteration of the `next()` method it calls `rowWriter.start()`.  Or does this work a little differently?

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/decoders/JsonMessageReader.java [hunk 2]
```diff
@@ -55,61 +47,78 @@
...
-        objectNode.put(KAFKA_PARTITION_ID.getFieldName(), record.partition());
-        objectNode.put(KAFKA_OFFSET.getFieldName(), record.offset());
-        objectNode.put(KAFKA_TIMESTAMP.getFieldName(), record.timestamp());
-        objectNode.put(KAFKA_MSG_KEY.getFieldName(), record.key() != null ? record.key().toString() : null);
-      } else {
-        throw new IOException("Unsupported node type: " + (jsonNode == null ? "NO CONTENT" : jsonNode.getNodeType()));
+      parseAndWrite(record, recordArray);
+    } catch (TokenIterator.RecoverableJsonException e) {
+      if (!readOptions.isSkipInvalidRecords()) {
+        throw e;
```
Comment (cgivre) [inline contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/decoders/JsonMessageReader.java:74]: Do you think we should throw a `UserException` here with an explanation and `errorContext`?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/JsonStructureOptions.java [hunk 3]
```diff
@@ -47,6 +47,13 @@
    */
   public boolean skipMalformedRecords;
 
+  /**
+   * This property works only when {@link #skipMalformedRecords} enabled.
+   * If true, {@link TokenIterator.RecoverableJsonException} will be populated for the case of
+   * malformed empty document, so it will be possible to handle this exception by caller.
+   */
+  public boolean skipMalformedDocument;
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/JsonStructureOptions.java:55]: Is there a system option for this?  Also is this documented anywhere?

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java [hunk 4]
```diff
@@ -17,131 +17,105 @@
...
+    try {
+      messageReader.close();
+    } catch (IOException e) {
+      logger.warn("Error closing Kafka message reader: {}", e.getMessage(), e);
+    }
   }
 
   @Override
   public String toString() {
     return "KafkaRecordReader[readOptions=" + readOptions
```
Comment (cgivre) [inline contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java:117]: Should we update this to use the `PlanStringBuilder`?  Here and elsewhere?

**Review by cgivre — CHANGES_REQUESTED** [review]

**Comment by cgivre** [comment]
@vvysotskyi  Thanks for this PR!   Does this PR address the single topic issue with the Kafka Plugin?

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java (hunk 1 above)
Comment (vvysotskyi) [reply contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java:79]: Yes, it works slightly differently, `rowWriter.start()` is called in `MessageReader.readMessage()` method after all checks below for batch size are passed.

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java (hunk 4 above)
Comment (vvysotskyi) [reply contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java]: Thanks, used it here.

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/decoders/JsonMessageReader.java (hunk 2 above)
Comment (vvysotskyi) [reply contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/decoders/JsonMessageReader.java]: Thanks, done

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/JsonStructureOptions.java (hunk 3 above)
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/JsonStructureOptions.java:55]: There is no option for this. It is for internal usage only.

**Review by vvysotskyi — COMMENTED** [review]
> @vvysotskyi Thanks for this PR! Does this PR address the single topic issue with the Kafka Plugin?
I wasn't able to reproduce this issue, looks like it was related to some older version of the Kafka driver.

File: contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java (hunk 1 above)
Comment (cgivre) [reply contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java:79]: Got it.  Thanks!

**Review by cgivre — APPROVED** [review]
LGTM +1

**Comment by cgivre** [comment]
@vvysotskyi  I approved this PR.  We should probably update the docs as well. @dzamo Can I tag you for that effort?

**Comment by luocooong** [comment]
A user asked Support confluent questions yesterday, update the docs for better. thanks.

**Pull request merged**